docs/architecture-decision-records/006-stage-awareness.md

```diff
@@ -0,0 +1,237 @@
+This means using a construct directly in a stage aware form becomes a little more difficult, however it reinforces the idea that patterns are the default entry point to the project.
+
+#### Score
+- :white_check_mark: Adherence to the best practice of producing a single, deterministic, artifact
+  - `cdk synth` will create a single CloudFormation template file
+- :white_check_mark: Adherence to the best practice of continuous delivery of infrastructure
+  - Riff-Raff supports the deployment of a single CloudFormation template file
+- :x: Support stacks that do not use `CODE` or `PROD`
+  - In order to support all possible stages [^1] in a typesafe way, it's likely the resulting code within GuCDK will be difficult to reason about
+  - GuCDK could enforce the use of `CODE`, `PROD` or `INFRA` only
+- :x: The simplicity of the API presented to users
+  - It would be possible for GuCDK to present a strongly typed API for any primitive type, however stronger types, such as [`InstanceType`][cdk-instance-type], will not be supported
+  - In this solution, `Stage`, being a parameter, is a [token][cdk-token], and cannot be used to perform logic. This is confusing and often causes runtime errors.
+- :white_check_mark: How short the feedback loop is
+  - The resulting CloudFormation template would look similar to how we write YAML CloudFormation
+  - The type system can help ensure we don't suffer from `FindInMap` issues listed above
+- :x: Ease of maintenance
+  - Although the type system could be used for guidance, in order to produce a stage aware component we'd either have to copy/paste some boilerplate, or introduce complex abstractions
+  - As this solution is directly against the recommendations from AWS, it might limit the evolution of GuCDK
+
+### 2. Provide a set of patterns and constructs with primitive types and the user provides a `Mapping` and `FindInMap` as needed
+- :white_check_mark: Adherence to the best practice of producing a single, deterministic, artifact
+  - `cdk synth` will create a single CloudFormation template file
+- :white_check_mark: Adherence to the best practice of continuous delivery of infrastructure
+  - Riff-Raff supports the deployment of a single CloudFormation template file
+- :white_check_mark: Support stacks that do not use `CODE` or `PROD`
+  - Users would be able to specify their own stages
+- :x: The simplicity of the API presented to users
+  - Users would have to create a Mapping themselves each time
+    - Creating [mappings within CDK][cdk-mapping] is quite verbose
+    - Goes against some aims of GuCDK which is to simplify infrastructure, making it easy to follow best practices
+  - Stronger, more type safe types such as [`InstanceType`][cdk-instance-type], will not be supported
+- :x: How short the feedback loop is
+  - Likely to be no faster than hand-writing a YAML template
```

Comment:
I'm not sure it's quite as bad as this suggests. A bit of work to create mappings yes, but with helpers it is pretty straightforward.

Reply:
I think we're suggesting GuCDK does not provide any helpers and user's would have to bring their own solution?